semantics/src/traversal/functions.rs

```diff
@@ -96,26 +94,26 @@ pub fn func_body(
         body,
     } = &def.node
     {
-        let name = name.node.to_string();
-        if let Some(ContractDef::Function {
-            is_public: _,
-            param_types: _,
-            return_type,
-            scope,
-        }) = contract_scope.borrow().def(name)
-        {
-            // If the return type is an empty tuple we do not have to validate any further
-            // at this point because both returning (explicit) or not returning (implicit
-            // return) are valid syntax.
-            // If the return type is anything else, we do need to ensure that all code paths
-            // return or revert.
-            if !return_type.is_empty_tuple() {
-                validate_all_paths_return_or_revert(&body)?
-            }
-
-            traverse_statements(Rc::clone(&scope), Rc::clone(&context), body)?;
+        let func_name = name.node.to_string();
+        let host_func_def = contract_scope
+            .borrow()
+            .function_def(func_name)
+            .expect(&format!(
+                "Failed to lookup function definition for {}",
+                &name.node
+            ));
```

Comment:
Calling `expect` here made more sense to me because the only reason why this could fail is a programming error on our end that needs to be fixed. And if that case happens, the message will give us a meaningful hint. It also seams that the old code would have returned `Ok(())` even if the `if let` would not have matched (because of an (unexpected) `None` case).